tsdb/head.go

```diff
@@ -1917,18 +1917,21 @@ func (h *headIndexReader) LabelValues(name string, matchers ...*labels.Matcher)
 
 // LabelNames returns all the unique label names present in the head
 // that are within the time range mint to maxt.
-func (h *headIndexReader) LabelNames() ([]string, error) {
-	h.head.symMtx.RLock()
+func (h *headIndexReader) LabelNames(matchers ...*labels.Matcher) ([]string, error) {
 	if h.maxt < h.head.MinTime() || h.mint > h.head.MaxTime() {
-		h.head.symMtx.RUnlock()
 		return []string{}, nil
 	}
 
-	labelNames := h.head.postings.LabelNames()
-	h.head.symMtx.RUnlock()
+	if len(matchers) == 0 {
+		h.head.symMtx.RLock()
```

Comment:
I kept this but I can't see why it's needed: `h.head.postings` is a `MemPostings`, and `MemPostings.LabelNames` [locks itself](http://github.com/prometheus/prometheus/blob/master/tsdb/index/postings.go#L84-L85), same happens for `LabelValues` above. the `symMtx` seems to be a mutex to be used to acces `h.head.symbols` but we're not accessing that in any way here so...

🤷